databases/Assignment2_answers.sql

```diff
@@ -0,0 +1,316 @@
+SELECT Product.Manufacturer
+FROM Product
+WHERE Computer.Speed < 3
+INNER JOIN Computer USING Model
+GROUP BY Product.Manufacturer
+HAVING COUNT(Product.Manufacturer) > 1;
+
+
+18. Which manufacturers sell the fastest computers?
+# general case where I do not know fastest speed
+SELECT DISTINCT Product.Manufacturer
+FROM Product
+INNER JOIN Computer USING (Model)
+WHERE Speed = max(Speed)
+# ^ verify if this is legit...it does not compile for me
+# instead doing:
+# WHERE Speed = (SELECT MAX(Speed) FROM Computer)
+# works, however this is basically a subquery.....so efficiency????? any way to not subquery?
+GROUP BY Manufacturer;
+
+
+
+# if I know the speed, then I can just do:
+SELECT Product.Manufacturer
+FROM Product
+INNER JOIN Computer
+WHERE Speed
+
+GROUP BY Product.Manufacturer
+HAVING
+
+
+19. Which manufacturers do not sell laptops?
+SELECT DISTINCT Manufacturer
+FROM Product
+EXCEPT
+SELECT DISTINCT Manufacturer
+FROM Product
+WHERE Type = laptop;
+
+# INCORRECT. does not account for case where manufacturer does laptop AND something else.
+SELECT DISTINCT Manufacturer
+FROM Product
+WHERE Type != laptop
+
+
+
+
+20. Which manufacturers offer at least 2 laptop models?
+SELECT Manufacturer
+FROM Product
+WHERE Type = laptop
+GROUP BY Manufacturer
+HAVING COUNT(Model) >= 2;
+
+
+21. Which manufacturers offer exactly 3 types of personal computers?
+SELECT Manufacturer
+FROM Product
+WHERE Type = pc
+GROUP BY Manufacturer
+HAVING COUNT(Model) = 3;
+
+22. You need to buy: # I can intersect this too, probably
+• a computer that has
+– at least 1024 MB of RAM,
+– a processor speed of 3 GHz or 3.5 GHz,
+– a HD of at least 160 GB, and
+• a color printer.
+Which manufacturers can offer you this combination and at what total price?
+
+
+with computers AS (
+    SELECT Computer.Model, Computer.Price, Product.Manufacturer
+    FROM Computer
+    INNER JOIN Product USING Model
+    WHERE RAM >= 1024 AND (Speed = 3 OR Speed = 3.5) AND HD >= 160
+),
+printers AS (
+    SELECT Computer.Model, Computer.Price, Product.Manufacturer
+    FROM Printer
+    INNER JOIN Product
+    WHERE Color = 'true'
+)
+SELECT computers.Manufacturer, (computers.Price + printers.Price) AS total_cost
+FROM computers
+CROSS JOIN printers
+WHERE computers.Manufacturer = printers.Manufacturers;
+
+
+
+# old version
+# WHY DOES THIS NOT WORK ?!?!?!?!?!
+--> because after your first inner join, you basically have one row left. then you try to inner join again,
+-- only this time there are no more commonalities, so you are left with an empty table to return.
+
+with computers AS (
+    SELECT Model, Price
+    FROM Computer
+    WHERE RAM > 1024 AND (Speed = 3 OR Speed = 3.5) AND HD >= 160
+),
+printers AS (
+    SELECT Model, Price
+    FROM Printer
+    WHERE Color = 'true'
+)
+SELECT Product.Manufacturer, (SUM(computers.Price) + SUM(printers.Price)) AS total_cost
+FROM Product
+INNER JOIN computers USING (Model)
+INNER JOIN printers USING (Model)
```

Comment:
You want to join on Manufacturer and not Model, because computers and printers have different model numbers